File: web/lab-record.js
/* Pure local laboratory records. Import validation is structural, not replay. */
(function expose(root, factory) {
  "use strict";
  const api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;
  if (root) root.VardeLabRecord = api;
})(typeof window === "object" ? window : null, function factory() {
  "use strict";
  const RULES = new Set(["line-breath", "gjerde-majority", "breath-connection", "junction-y", "junction-six", "junction-planted", "junction-passage"]);
  const BUILD = {"junction-y": ["construct", 2], "junction-six": ["construct", 1], "junction-planted": ["plant", 2], "junction-passage": ["construct", 3]};
  const TOP = ["format", "version", "source", "session_id", "rules", "board_size", "catalog_version", "initial_seats", "status", "actions", "final_score"];
  const AFTER = ["to_move", "actor_color", "actor_seat", "seats", "end_acceptances", "finished", "accepted", "resumption_used", "consecutive_passes", "swap_available", "moves_played", "placements_played", "constructions_played", "topology", "score", "original_control"];
  const COLORS = ["B", "W"], SEATS = ["S1", "S2"];
  const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
  const copy = (value) => JSON.parse(JSON.stringify(value));
  const canonical = (value) => {
    if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
    if (value && typeof value === "object") return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`).join(",")}}`;
    return JSON.stringify(value);
  };
  const equal = (left, right) => canonical(left) === canonical(right);
  const coordOrder = (a, b) => a[0] - b[0] || a[1] - b[1] || (a[2] ?? 0) - (b[2] ?? 0);
  const fail = (message) => { throw new Error(message); };
  function fields(value, expected, label) {
    if (!value || typeof value !== "object" || Array.isArray(value)
        || !equal(Object.keys(value).sort(), [...expected].sort())) fail(`Invalid ${label} fields`);
  }
  function integer(value, label, minimum = 0) {
    if (!Number.isSafeInteger(value) || value < minimum) fail(`Invalid ${label} integer`);
  }
  function point(value) {
    if (!Array.isArray(value) || value.length !== 2) fail("Invalid coordinate");
    value.forEach((coordinate) => integer(coordinate, "coordinate", -Number.MAX_SAFE_INTEGER));
  }
  function score(value, label) {
    fields(value, COLORS, label);
    Object.values(value).forEach((number) => integer(number, label, label === "score" ? -Number.MAX_SAFE_INTEGER : 0));
  }
  function seats(value) {
    fields(value, COLORS, "seats");
    if (!equal(Object.values(value).sort(), SEATS)) fail("Only bijective S1/S2 neutral seats are allowed");
  }
  function actionShape(value, rules) {
    if (!value || typeof value !== "object" || typeof value.action !== "string") fail("Invalid action");
    const kind = value.action;
    if (kind === "construct" || kind === "plant") {
      fields(value, ["action", "face", "orientation"], "construction");
      point(value.face);
      integer(value.orientation, "orientation");
      const [expected, count] = BUILD[rules] || [null, 0];
      if (kind !== expected || value.orientation >= count) fail("Action or orientation unavailable for these rules");
    } else if (kind === "play") {
      fields(value, ["action", "point"], "placement"); point(value.point);
    } else if (["pass", "swap", "accept", "resume"].includes(kind)) {
      fields(value, ["action"], "administrative action");
    } else fail("Unsupported lab record action");
    return kind;
  }
  function afterShape(value, rules) {
    fields(value, AFTER, "after projection"); seats(value.seats);
    if (!COLORS.includes(value.to_move)) fail("Invalid next color");
    ["finished", "accepted", "resumption_used", "swap_available"].forEach((key) => {
      if (typeof value[key] !== "boolean") fail(`Invalid ${key} flag`);
    });
    ["consecutive_passes", "moves_played", "placements_played", "constructions_played"].forEach((key) => integer(value[key], key));
    if (value.consecutive_passes > 2 || value.finished !== (value.consecutive_passes === 2)) fail("Inconsistent ending/pass projection");
    const accepts = value.end_acceptances;
    if (!Array.isArray(accepts) || accepts.some((seat) => !SEATS.includes(seat))
        || !equal(accepts, [...new Set(accepts)].sort())) fail("Invalid neutral acceptance list");
    if (value.accepted) {
      if (!value.finished || value.actor_color !== null || value.actor_seat !== null) fail("Accepted state must have no next actor");
      if (accepts.length !== (value.resumption_used ? 1 : 2)) fail("Invalid terminal acceptance count");
    } else {
      if (!COLORS.includes(value.actor_color) || value.actor_seat !== value.seats[value.actor_color]) fail("Invalid acting neutral seat");
      if (!value.finished && (accepts.length || value.actor_color !== value.to_move)) fail("Invalid live actor or acceptance");
      if (value.finished && accepts.length > 1) fail("Too many pending acceptances");
    }
    if (!Array.isArray(value.topology)) fail("Invalid topology");
    const count = (BUILD[rules] || [null, 0])[1], faces = new Set();
    value.topology.forEach((row) => {
      if (!Array.isArray(row) || row.length !== 3) fail("Invalid topology row");
      point(row.slice(0, 2)); integer(row[2], "orientation");
      const face = JSON.stringify(row.slice(0, 2));
      if (row[2] >= count || faces.has(face)) fail("Invalid or duplicate topology face");
      faces.add(face);
    });
    if (!equal(value.topology, [...value.topology].sort(coordOrder)) || value.topology.length !== value.constructions_played) fail("Noncanonical topology or construction count");
    score(value.score, "score"); score(value.original_control, "original control");
  }
  function capturesShape(value) {
    fields(value, ["original", "junction", "waves"], "captures");
    integer(value.original, "original captures"); integer(value.junction, "junction captures");
    if (!Array.isArray(value.waves)) fail("Invalid capture waves");
    const seen = new Set();
    value.waves.forEach((wave) => {
      if (!Array.isArray(wave) || !wave.length) fail("Invalid empty capture wave");
      wave.forEach((coordinate) => {
        point(coordinate);
        const key = JSON.stringify(coordinate);
        if (seen.has(key)) fail("Duplicate captured coordinate");
        seen.add(key);
      });
      if (!equal(wave, [...wave].sort(coordOrder))) fail("Capture wave points must be canonical");
    });
    if (seen.size !== value.original + value.junction) fail("Capture totals and wave coordinates differ");
  }
  function validateLabRecordShape(record) {
    fields(record, TOP, "lab record");
    if (record.format !== "varde-lab-playtest" || record.version !== 2) fail("Unsupported lab record");
    if (record.source !== "browser-local-hotseat" || typeof record.session_id !== "string" || !UUID.test(record.session_id)) fail("Invalid local source or UUID");
    fields(record.rules, ["id", "revision"], "rules");
    if (!RULES.has(record.rules.id) || record.rules.revision !== "0.1") fail("Unsupported lab rules/revision");
    integer(record.board_size, "board size"); integer(record.catalog_version, "catalog version", 1);
    if (![3, 4, 5, 6].includes(record.board_size)) fail("Unsupported lab board size");
    seats(record.initial_seats);
    if (!equal(record.initial_seats, {B: "S1", W: "S2"})) fail("Initial neutral seats must be B:S1/W:S2");
    if (!["active", "complete"].includes(record.status) || !Array.isArray(record.actions)) fail("Invalid record status or action list");
    let previous = null, currentSeats = record.initial_seats;
    const counters = {moves_played: 0, placements_played: 0, constructions_played: 0};
    record.actions.forEach((row, index) => {
      fields(row, ["index", "action", "actor", "elapsed_ms", "captures", "after"], "action row");
      integer(row.index, "action index");
      if (row.index !== index) fail("Nonsequential action index");
      const kind = actionShape(row.action, record.rules.id);
      fields(row.actor, ["seat", "color"], "actor");
      const actor = previous ? {seat: previous.actor_seat, color: previous.actor_color} : {seat: "S1", color: "B"};
      if (!equal(row.actor, actor) || actor.seat === null) fail("Action actor differs from recorded predecessor");
      if (index === 0 && kind !== "play") fail("First action must be original placement");
      integer(row.elapsed_ms, "elapsed time"); capturesShape(row.captures);
      if (!["play", "plant"].includes(kind) && !equal(row.captures, {original: 0, junction: 0, waves: []})) fail("Administrative or empty construction action cannot capture");
      const after = row.after;
      afterShape(after, record.rules.id);
      const expectedSeats = kind === "swap" ? {B: currentSeats.W, W: currentSeats.B} : currentSeats;
      if (!equal(after.seats, expectedSeats)) fail("Unexplained neutral seat reassignment");
      counters.moves_played += Number(["play", "plant", "construct", "pass"].includes(kind));
      counters.placements_played += Number(["play", "plant"].includes(kind));
      counters.constructions_played += Number(["construct", "plant"].includes(kind));
      Object.keys(counters).forEach((key) => { if (after[key] !== counters[key]) fail("Action/counter structure differs"); });
      currentSeats = after.seats; previous = after;
    });
    const complete = previous !== null && previous.accepted;
    if ((record.status === "complete") !== complete) fail("Completion requires accepted terminal");
    if (complete) {
      score(record.final_score, "score");
      if (!equal(record.final_score, previous.score)) fail("Final score differs from last accepted score");
    } else if (record.final_score !== null) fail("Active record cannot have final score");
    return copy(record);
  }
  function identities(view) {
    if (!view || view.experimental !== true || view.match?.mode !== "hotseat" || !RULES.has(view.rules)) fail("Recording requires a hotseat laboratory game");
    const ids = {};
    COLORS.forEach((color) => {
      const seat = view.match.seats?.[color];
      if (seat?.kind !== "human" || typeof seat.identity !== "string" || !seat.identity) fail("Recording requires two human seats");
      ids[color] = seat.identity;
    });
    if (ids.B === ids.W || (!view.accepted && view.actor_seat !== ids[view.actor_color])) fail("Invalid actual actor identity");
    return ids;
  }
  function projection(view, neutral) {
    const ids = identities(view);
    const translate = (identity) => {
      const color = COLORS.find((item) => ids[item] === identity);
      if (!color) fail("Unknown current acceptance identity");
      return neutral[color];
    };
    const output = {};
    AFTER.forEach((key) => {
      if (!["seats", "end_acceptances", "actor_seat"].includes(key)) output[key] = copy(view[key]);
    });
    output.seats = copy(neutral);
    output.actor_seat = view.actor_seat === null ? null : translate(view.actor_seat);
    if (!Array.isArray(view.match.end_acceptances)) fail("Missing seat acceptances");
    output.end_acceptances = view.match.end_acceptances.map(translate).sort();
    afterShape(output, view.rules);
    return output;
  }
  function createLabRecord(view, {sessionId, catalogVersion}) {
    identities(view);
    if (/^[0-9a-f]{32}$/.test(sessionId)) sessionId = `${sessionId.slice(0, 8)}-${sessionId.slice(8, 12)}-${sessionId.slice(12, 16)}-${sessionId.slice(16, 20)}-${sessionId.slice(20)}`;
    const initial = projection(view, {B: "S1", W: "S2"});
    if (initial.moves_played || initial.placements_played || initial.constructions_played || initial.topology.length
        || initial.finished || initial.accepted || initial.resumption_used || initial.swap_available
        || initial.consecutive_passes || initial.actor_color !== "B" || initial.end_acceptances.length
        || !Array.isArray(view.points) || !view.points.length || view.points.some((point) => !Array.isArray(point.stack) || point.stack.length)) fail("Start requires a fresh game before the opening");
    return validateLabRecordShape({
      format: "varde-lab-playtest", version: 2, source: "browser-local-hotseat", session_id: sessionId,
      rules: {id: view.rules, revision: view.rules_revision}, board_size: view.n,
      catalog_version: catalogVersion, initial_seats: {B: "S1", W: "S2"},
      status: "active", actions: [], final_score: null,
    });
  }
  function appendLabAction(record, {action, before, after, elapsedMs}) {
    const next = validateLabRecordShape(record);
    if (next.status === "complete") fail("Accepted record cannot continue");
    [before, after].forEach((view) => {
      identities(view);
      if (view.rules !== next.rules.id || view.rules_revision !== next.rules.revision || view.n !== next.board_size) fail("Displayed game differs from record");
    });
    const kind = actionShape(action, next.rules.id);
    const previous = next.actions.at(-1)?.after;
    const neutral = previous?.seats || next.initial_seats;
    const projectedBefore = projection(before, neutral);
    if (previous && !equal(previous, projectedBefore)) fail("Displayed game differs from recorded prefix");
    if (!previous) createLabRecord(before, {sessionId: next.session_id, catalogVersion: next.catalog_version});
    const idsBefore = identities(before), idsAfter = identities(after);
    const expectedIds = kind === "swap" ? {B: idsBefore.W, W: idsBefore.B} : idsBefore;
    if (!equal(idsAfter, expectedIds)) fail("Actual identities changed outside the recorded action");
    const nextNeutral = kind === "swap" ? {B: neutral.W, W: neutral.B} : neutral;
    const projectedAfter = projection(after, nextNeutral);
    if (before.legal_actions !== undefined
        && (!Array.isArray(before.legal_actions) || !before.legal_actions.some((item) => equal(item, action)))) fail("Action was not in the authoritative legal action list");
    integer(elapsedMs, "elapsed time");
    const afterPoints = new Map(after.points.map((point) => [JSON.stringify(point.coord), point]));
    const captured = {original: 0, junction: 0, waves: []};
    const removed = new Set();
    before.points.forEach((point) => {
      if (typeof point.original !== "boolean") fail("Missing original-point classification");
      const key = JSON.stringify(point.coord), advanced = afterPoints.get(key);
      if (!advanced) fail("Permanent topology point disappeared");
      if (point.stack.length && !advanced.stack.length) {
        captured[point.original ? "original" : "junction"] += 1;
        removed.add(key);
      }
    });
    if (["play", "plant"].includes(kind)) {
      if (!Array.isArray(after.capture_waves)) fail("Missing capture wave telemetry");
      captured.waves = after.capture_waves.map((wave) => copy(wave).sort(coordOrder));
    }
    capturesShape(captured);
    const wavePoints = captured.waves.flat().map((point) => JSON.stringify(point));
    if (!equal([...removed].sort(), wavePoints.sort())) fail("Capture waves differ from removed stones");
    next.actions.push({index: next.actions.length, action: copy(action),
      actor: {seat: projectedBefore.actor_seat, color: before.actor_color}, elapsed_ms: elapsedMs,
      captures: captured, after: projectedAfter});
    next.status = after.accepted ? "complete" : "active";
    next.final_score = after.accepted ? copy(after.score) : null;
    return validateLabRecordShape(next);
  }
  return Object.freeze({createLabRecord, appendLabAction, validateLabRecordShape});
});

File: research/harness/lab_record.py
"""Strict local lab-record shapes and mechanical shared-action replay.

Version-1 human-study records are deliberately not accepted or modified here.
No server, opponent or Personal model is imported. CLI reads one local file and
prints a separate verification report; it never stores or submits a record.
"""

import argparse
from copy import deepcopy
from dataclasses import dataclass
import hashlib
import json
import math
from pathlib import Path
import re
import sys

ENGINE_ROOT = Path(__file__).resolve().parents[2] / "engine"
if str(ENGINE_ROOT) not in sys.path:
    sys.path.insert(0, str(ENGINE_ROOT))

from actions import RulesAction, RulesState, apply_action  # noqa: E402
from game_factory import new_game  # noqa: E402
from lab_spec import LAB_REGISTRY  # noqa: E402
from varde import Illegal  # noqa: E402


FORMAT = "varde-lab-playtest"
VERSION = 2
MAX_INTEGER = 2**53 - 1
SEATS = {"S1", "S2"}
COLORS = {"B", "W"}
TOP_FIELDS = {"format", "version", "source", "session_id", "rules", "board_size",
              "catalog_version", "initial_seats", "status", "actions", "final_score"}
ROW_FIELDS = {"index", "action", "actor", "elapsed_ms", "captures", "after"}
AFTER_FIELDS = {"to_move", "actor_color", "actor_seat", "seats", "end_acceptances",
                "finished", "accepted", "resumption_used", "consecutive_passes", "swap_available",
                "moves_played", "placements_played", "constructions_played", "topology", "score", "original_control"}
CONSTRUCTION = {"junction-y": ("construct", 2), "junction-six": ("construct", 1),
                "junction-planted": ("plant", 2), "junction-passage": ("construct", 3)}
UUID = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")


def _detached_json_numbers(value):
    """Match parsed JS Number semantics without relaxing engine action parsing."""
    if type(value) is float:
        if not math.isfinite(value) or not value.is_integer() or abs(value) > MAX_INTEGER:
            raise ValueError("record numbers must be finite safe integers")
        return int(value)
    if type(value) is dict:
        return {key: _detached_json_numbers(item) for key, item in value.items()}
    if type(value) is list:
        return [_detached_json_numbers(item) for item in value]
    # Booleans remain booleans: an integer field must still reject them.
    return deepcopy(value)


def _fields(value, fields, label):
    if type(value) is not dict or set(value) != fields:
        raise ValueError(f"invalid {label} fields")


def _integer(value, label, *, minimum=0):
    if type(value) is not int or not minimum <= value <= MAX_INTEGER:
        raise ValueError(f"invalid {label} integer")


def _point(value):
    if type(value) is not list or len(value) != 2:
        raise ValueError("invalid coordinate")
    for coordinate in value:
        _integer(coordinate, "coordinate", minimum=-MAX_INTEGER)


def _score(value, label):
    _fields(value, COLORS, label)
    for score in value.values():
        _integer(score, label, minimum=-MAX_INTEGER if label == "score" else 0)


def _seats(value):
    _fields(value, COLORS, "seats")
    if any(type(seat) is not str for seat in value.values()) or set(value.values()) != SEATS:
        raise ValueError("only bijective S1/S2 neutral seats are allowed")


def _action(value, rules):
    parsed = RulesAction.from_dict(value)
    kind = parsed.kind
    if kind in ("construct", "plant"):
        expected, count = CONSTRUCTION.get(rules, (None, 0))
        if kind != expected or parsed.orientation >= count:
            raise ValueError("action or orientation unavailable for these rules")
    elif kind not in ("play", "pass", "swap", "accept", "resume"):
        raise ValueError("unsupported lab record action")
    if parsed.point is not None:
        _point(value["face" if kind in ("construct", "plant") else "point"])
    return parsed


def _after(value, rules):
    _fields(value, AFTER_FIELDS, "after projection")
    _seats(value["seats"])
    if type(value["to_move"]) is not str or value["to_move"] not in COLORS:
        raise ValueError("invalid next color")
    for key in ("finished", "accepted", "resumption_used", "swap_available"):
        if type(value[key]) is not bool:
            raise ValueError(f"invalid {key} flag")
    for key in ("consecutive_passes", "moves_played", "placements_played", "constructions_played"):
        _integer(value[key], key)
    if value["consecutive_passes"] > 2 or value["finished"] != (value["consecutive_passes"] == 2):
        raise ValueError("inconsistent ending/pass projection")
    accepts = value["end_acceptances"]
    if (type(accepts) is not list or any(type(seat) is not str or seat not in SEATS for seat in accepts)
            or accepts != sorted(set(accepts))):
        raise ValueError("invalid neutral acceptance list")
    if value["accepted"]:
        if not value["finished"] or value["actor_color"] is not None or value["actor_seat"] is not None:
            raise ValueError("accepted state must have no next actor")
        if len(accepts) != (1 if value["resumption_used"] else 2):
            raise ValueError("invalid terminal acceptance count")
    else:
        color = value["actor_color"]
        if type(color) is not str or color not in COLORS or value["actor_seat"] != value["seats"][color]:
            raise ValueError("invalid acting neutral seat")
        if not value["finished"] and (accepts or color != value["to_move"]):
            raise ValueError("invalid live actor or acceptance")
        if value["finished"] and len(accepts) > 1:
            raise ValueError("too many pending acceptances")
    topology = value["topology"]
    if type(topology) is not list:
        raise ValueError("invalid topology")
    count = CONSTRUCTION.get(rules, (None, 0))[1]
    faces = set()
    for row in topology:
        if type(row) is not list or len(row) != 3:
            raise ValueError("invalid topology row")
        _point(row[:2])
        _integer(row[2], "orientation")
        if row[2] >= count or tuple(row[:2]) in faces:
            raise ValueError("invalid or duplicate topology face")
        faces.add(tuple(row[:2]))
    if topology != sorted(topology) or len(topology) != value["constructions_played"]:
        raise ValueError("noncanonical topology or construction count")
    _score(value["score"], "score")
    _score(value["original_control"], "original control")


def _captures(value):
    _fields(value, {"original", "junction", "waves"}, "captures")
    _integer(value["original"], "original captures")
    _integer(value["junction"], "junction captures")
    if type(value["waves"]) is not list:
        raise ValueError("invalid capture waves")
    seen = set()
    for wave in value["waves"]:
        if type(wave) is not list or not wave:
            raise ValueError("invalid empty capture wave")
        for point in wave:
            _point(point)
            if tuple(point) in seen:
                raise ValueError("duplicate captured coordinate")
            seen.add(tuple(point))
        if wave != sorted(wave):
            raise ValueError("capture wave points must be canonical")
    if len(seen) != value["original"] + value["junction"]:
        raise ValueError("capture totals and wave coordinates differ")


def validate_lab_record_shape(record):
    """Return detached strict structural data, not a mechanical certificate."""
    record = _detached_json_numbers(record)
    _fields(record, TOP_FIELDS, "lab record")
    if record["format"] != FORMAT or type(record["version"]) is not int or record["version"] != VERSION:
        raise ValueError("unsupported lab record")
    if record["source"] != "browser-local-hotseat" or not isinstance(record["session_id"], str) or not UUID.fullmatch(record["session_id"]):
        raise ValueError("invalid local source or UUID")
    _fields(record["rules"], {"id", "revision"}, "rules")
    rules = record["rules"]["id"]
    if not isinstance(rules, str) or rules not in LAB_REGISTRY or record["rules"]["revision"] != "0.1":
        raise ValueError("unsupported lab rules/revision")
    _integer(record["board_size"], "board size")
    if record["board_size"] not in (3, 4, 5, 6):
        raise ValueError("unsupported lab board size")
    _integer(record["catalog_version"], "catalog version", minimum=1)
    _seats(record["initial_seats"])
    if record["initial_seats"] != {"B": "S1", "W": "S2"}:
        raise ValueError("initial neutral seats must be B:S1/W:S2")
    if record["status"] not in ("active", "complete") or type(record["actions"]) is not list:
        raise ValueError("invalid record status or action list")
    previous = None
    seats = record["initial_seats"]
    counters = {"moves_played": 0, "placements_played": 0, "constructions_played": 0}
    for index, row in enumerate(record["actions"]):
        _fields(row, ROW_FIELDS, "action row")
        _integer(row["index"], "action index")
        if row["index"] != index:
            raise ValueError("nonsequential action index")
        action = _action(row["action"], rules)
        _fields(row["actor"], {"seat", "color"}, "actor")
        expected_actor = {"seat": "S1", "color": "B"} if previous is None else {
            "seat": previous["actor_seat"], "color": previous["actor_color"]}
        if row["actor"] != expected_actor or expected_actor["seat"] is None:
            raise ValueError("action actor differs from recorded predecessor")
        if index == 0 and action.kind != "play":
            raise ValueError("first action must be original placement")
        _integer(row["elapsed_ms"], "elapsed time")
        _captures(row["captures"])
        if action.kind not in ("play", "plant") and row["captures"] != {"original": 0, "junction": 0, "waves": []}:
            raise ValueError("administrative or empty construction action cannot capture")
        after = row["after"]
        _after(after, rules)
        expected_seats = {"B": seats["W"], "W": seats["B"]} if action.kind == "swap" else seats
        if after["seats"] != expected_seats:
            raise ValueError("unexplained neutral seat reassignment")
        deltas = {"moves_played": int(action.kind in ("play", "plant", "construct", "pass")),
                  "placements_played": int(action.kind in ("play", "plant")),
                  "constructions_played": int(action.kind in ("construct", "plant"))}
        for key in counters:
            counters[key] += deltas[key]
            if after[key] != counters[key]:
                raise ValueError("action/counter structure differs")
        seats, previous = after["seats"], after
    complete = previous is not None and previous["accepted"]
    if (record["status"] == "complete") != complete:
        raise ValueError("completion requires accepted terminal")
    if complete:
        _score(record["final_score"], "score")
        if record["final_score"] != previous["score"]:
            raise ValueError("final score differs from last accepted score")
    elif record["final_score"] is not None:
        raise ValueError("active record cannot have final score")
    return deepcopy(record)


def _projection(state):
    game = state.game
    return {"to_move": game.to_move, "actor_color": state.actor_color, "actor_seat": state.actor_seat,
            "seats": dict(state.seats), "end_acceptances": sorted(state.end_acceptances),
            "finished": game.finished, "accepted": state.accepted, "resumption_used": game.resumption_used,
            "consecutive_passes": game.consecutive_passes, "swap_available": game.swap_available,
            "moves_played": game.moves_played, "placements_played": game.placements_played,
            "constructions_played": game.constructions_played,
            "topology": [list(row) for row in game.topology], "score": game.score(),
            "original_control": game.original_control_count()}


def _capture_projection(before, after, action):
    original = frozenset(getattr(before.game.board, "original_points", before.game.board.points))
    removed = {point for point, stack in before.game.state.items() if stack and not after.game.state.get(point)}
    waves = after.game.last_capture_waves if action.kind in ("play", "plant") else []
    return {"original": len(removed & original), "junction": len(removed - original),
            "waves": [[list(point) for point in sorted(wave)] for wave in waves]}


@dataclass(frozen=True)
class LabRecordReplay:
    action_count: int
    accepted: bool
    replay_hash: str
    mechanically_verified: bool = True

    def to_dict(self):
        return {"format": "varde-lab-record-replay", "version": 1,
                "mechanically_verified": self.mechanically_verified, "action_count": self.action_count,
                "accepted": self.accepted, "replay_hash": self.replay_hash,
                "claim_limit": "mechanical replay only; not a qualified human study or game-quality finding"}


def replay_lab_record(record):
    record = validate_lab_record_shape(record)
    game = new_game(record["board_size"], rules=record["rules"]["id"], experimental=True)
    game.players = {"B": "S1", "W": "S2"}
    state = RulesState(game, seats=dict(record["initial_seats"]))
    for row in record["actions"]:
        index = row["index"]
        if row["actor"] != {"seat": state.actor_seat, "color": state.actor_color}:
            raise ValueError(f"action {index}: actor mismatch")
        action = RulesAction.from_dict(row["action"])
        before = state
        try:
            state = apply_action(state, action)
        except (Illegal, ValueError, RuntimeError) as error:
            raise ValueError(f"action {index}: illegal recorded transition") from error
        if row["after"] != _projection(state):
            raise ValueError(f"action {index}: replayed state/telemetry mismatch")
        if row["captures"] != _capture_projection(before, state, action):
            raise ValueError(f"action {index}: replayed captures mismatch")
    semantic = {"rules": record["rules"], "board_size": record["board_size"],
                "actions": [row["action"] for row in record["actions"]], "state": state.to_dict()}
    digest = hashlib.sha256(json.dumps(semantic, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()).hexdigest()
    return LabRecordReplay(len(record["actions"]), state.accepted, digest)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", type=Path, help="local version-2 lab record JSON")
    args = parser.parse_args()
    try:
        with args.input.open() as stream:
            record = json.load(stream)
        result = replay_lab_record(record)
    except (OSError, ValueError) as error:
        parser.error(str(error))
    print(json.dumps(result.to_dict(), sort_keys=True))


if __name__ == "__main__":
    main()
